SourceX/qol.cpp

```diff
@@ -216,4 +216,157 @@ void AutoGoldPickup(int pnum)
 	}
 }
 
+struct drawingQueue {
+	int itemID, row, col, x, y, width, height, color;
+	std::string text;
+	drawingQueue(int itemID, int row, int col, int x, int y, int width, int height, int color, std::string text)
+	    : itemID(itemID)
+	    , row(row)
+	    , col(col)
+	    , x(x)
+	    , y(y)
+	    , width(width)
+	    , height(height)
+	    , color(color)
+	    , text(text)
+	{
+	}
+};
+
+std::vector<drawingQueue> drawQ;
+int drawMinX;
+int drawMaxX;
+bool altPressed = false;
+int generatingLabelsState = 0; // 0 - waiting for the game to start, 1 = started generating, 2 = finished generating
+bool isLabelHighlighted = false;
+int labelCenterOffsets[ITEMTYPES];
+
+void AltPressed(bool pressed)
+{
+	altPressed = pressed;
+}
+
+bool IsGeneratingLabels()
+{
+	return generatingLabelsState == 1;
+}
+
+bool IsLabelHighlighted()
+{
+	return isLabelHighlighted;
+}
+
+void UpdateLabelOffsets(CelOutputBuffer out, BYTE *dst, int width)
+{
+	int xval = (dst - out.begin()) % out.pitch();
+	if (xval < drawMinX)
+		drawMinX = xval;
+	xval += width;
+	if (xval > drawMaxX)
+		drawMaxX = xval;
+}
+
+void GenerateLabelOffsets(CelOutputBuffer out)
+{
+	if (generatingLabelsState == 2)
+		return;
+	generatingLabelsState = 1;
+	int itemTypes = gbIsHellfire ? ITEMTYPES : 35;
+	for (int i = 0; i < itemTypes; i++) {
+		drawMinX = gnScreenWidth;
+		drawMaxX = 0;
+		CelClippedDrawTo(out, out.pitch() / 2 - 16, 351, itemanims[i], ItemAnimLs[i], 96);
+		labelCenterOffsets[i] = drawMinX - out.pitch() / 2 + (drawMaxX - drawMinX) / 2;
+	}
+	generatingLabelsState = 2;
+}
+
+void AddItemToDrawQueue(int x, int y, int id)
+{
+	if (sgOptions.Gameplay.bAltHighlight == 0 || (sgOptions.Gameplay.bAltHighlight == 1 && !altPressed) || (sgOptions.Gameplay.bAltHighlight == 2 && altPressed))
+		return;
+	ItemStruct *it = &item[id];
+
+	char textOnGround[64];
+	if (it->_itype == ITYPE_GOLD)
+		sprintf(textOnGround, "%i gold", it->_ivalue);
+	else
+		sprintf(textOnGround, "%s", it->_iIdentified ? it->_iIName : it->_iName);
+
+	int nameWidth = GetTextWidth((char *)textOnGround);
+	x += labelCenterOffsets[ItemCAnimTbl[it->_iCurs]];
+	y -= TILE_HEIGHT;
+	if (!zoomflag) {
+		x <<= 1;
+		y <<= 1;
+	}
+	x -= nameWidth >> 1;
+	char clr = COL_WHITE;

```

Comment:
Minor: this attribute could be a text_color to avoid the cast later.

Reply:
I tried but compiler cried so I just stopped caring lol

Reply:
:sob:

Reply:
I did that in a branch somewhere but I had to change the "Drawing Item" struct too.

Reply:
oh well, worked this time 🗡️